hivemind/client/averaging/__init__.py

```diff
@@ -279,6 +289,134 @@ async def rpc_aggregate_part(self, stream: AsyncIterator[averaging_pb2.Averaging
         async for message in group.rpc_aggregate_part(achain(aiter(request), stream), context):
             yield message
 
+    async def _declare_for_download_periodically(self):
+        download_key = f'{self._matchmaking.group_key_manager.prefix}.all_averagers'
+        while True:
+            asyncio.create_task(asyncio.wait_for(self.dht.store(
+                download_key, subkey=self.endpoint, value=self.last_updated,
+                expiration_time=get_dht_time() + self._matchmaking.averaging_expiration, return_future=True),
+                timeout=self._matchmaking.averaging_expiration))
+            await asyncio.sleep(self._matchmaking.averaging_expiration)
+
+    async def rpc_download_state(self, request: averaging_pb2.DownloadRequest, context: grpc.ServicerContext
+                                 ) -> AsyncIterator[averaging_pb2.DownloadData]:
+        """ a newcomer requests us to send over our current trainer state for his initialization """
+        chunk_size_bytes = self.matchmaking_kwargs.get('chunk_size_bytes', 2 ** 16)
+        metadata, tensors = await self._get_current_state_from_host_process()
+
+        for tensor in tensors:
+            for part in split_for_streaming(serialize_torch_tensor(tensor), chunk_size_bytes):
+                if metadata is not None:
+                    yield averaging_pb2.DownloadData(tensor_part=part, metadata=metadata)
+                    metadata = None
+                else:
+                    yield averaging_pb2.DownloadData(tensor_part=part)
```

Comment:
Looks kinda hacky if your goal is to send the metadata only with the first part of the first tensor. Can't we devise a better protocol here? (for instance, send only the metadata first, then send only the tensor parts)